src/unix/bsd/freebsdlike/freebsd/mod.rs

```diff
@@ -1202,6 +1229,20 @@ s_no_extra_traits! {
     }
 }
 
+impl ::core::ops::Deref for sigevent {
+    type Target = sigevent_0_2_126;
+
+    fn deref(&self) -> &Self::Target {
+        unsafe { &*(self as *const Self as *const sigevent_0_2_126) }
```

Comment:
This Deref impl exposes UB to safe code on freebsd.

```rust
fn main() {
    let sigevent = libc::sigevent {
        sigev_notify: 0,
        sigev_signo: 0,
        sigev_value: libc::sigval {
            sival_ptr: 0 as *mut _,
        },
        _sigev_un: libc::__c_anonymous_sigev_un {
            _kevent_flags: 0,
        },
    };
    let _val = sigevent.sigev_notify_thread_id;
}
```

Tested in miri using https://github.com/rust-lang/miri/pull/2221:

```console
$ cargo miri run --target x86_64-unknown-freebsd

error: Undefined Behavior: type validation failed: encountered uninitialized bytes, but expected initialized bytes
  --> src/main.rs:12:16
   |
12 |     let _val = sigevent.sigev_notify_thread_id;
   |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ type validation failed: encountered uninitialized bytes, but expected initialized bytes
   |
   = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
   = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
           
   = note: inside `main` at src/main.rs:12:16
```

@Amanieu I thought this exact unsoundness issue came up in a recent T-libs-api meeting. Did we end up deciding that this is "okay enough" as a tradeoff? I didn't see it discussed explicitly here on the PR.

![Screenshot from 2022-06-20 21-58-56](https://user-images.githubusercontent.com/1940490/174720033-29856fb0-5996-4cc1-a476-9fbfe58546ec.png)

Reply:
But that field will never be uninitialized, because no program should ever use both `_sigev_un` and `sigev_notify_thread_id`.  The latter only exists to provide backwards-compatibility with the pre-union code.  So anybody accessing that field should already have initialized the entire structure.

Reply:
I get that what you said is what programs _should_ do. If the libc crate ensures that programs do as you described, then that's great. If the **programmer** needs to ensure they do as you described or else they get memory unsafety (which is what is currently implemented in this PR) then we need to make them write `unsafe` somewhere.

Reply:
The problem is that the old, non-union version of `sigevent` is already in the wild.  And since it isn't `unsafe`, there simply isn't any way for libc to fix the definition of `sigevent`, be backwards-compatible, *and* be `unsafe`.  The deref compromise suggested by @Amanieu is IMHO the best compromise.
FWIW there is exactly one currently-maintained crate on crates.io that uses the old `sigev_notify_thread_id` field, and I'm a maintainer of that crate, so I can fix any resulting bugs.  It's my intention to release a new version of Nix as soon as libc releases a new version with this PR.  However, the backwards-compatibility is still useful for users of older Nix versions.

Reply:
> there simply isn't any way for libc to fix the definition of `sigevent`, be backwards-compatible, *and* be `unsafe`

Before this PR, libc::sigevent cannot be constructed in safe code because it has private padding fields. After this PR it can be constructed as shown above, which is a soundness bug. Preserving the property that it cannot be constructed in safe code is not a breaking change. The safety requirement on constructing a sigevent can be that you never expose it in a way that would mismatch the sigev_notify with the active variant of _sigev_un.

You can use #\[non_exhaustive\] if libc uses that these days, or just a zero sized private field with appropriate comment.